[PLUGIN-1735] Fix read timeout issue during GCS Bucket  copy/move actions for large files

dj-smart opened the pull request:
[JIRA
](https://cdap.atlassian.net/browse/PLUGIN-1735
)
## Introduction:
- While copying large files using the **GCS Copy** plugin, `Read Timed out` error was occuring as Storage Client was configured with the default timeout of 20 seconds: https://cloud.google.com/storage/docs/retry-strategy#client-libraries

## Fix
- A property **readTimeout** has been introduced for GCS Copy/Move Plugins that enables user to provide the timeout based on their requirements depending on the file size that is being transferred.
- The Retry Settings for the Storage Client are configured to use this property using [this method](https://cloud.google.com/java/docs/reference/google-cloud-core/latest/com.google.cloud.http.HttpTransportOptions.Builder#com_google_cloud_http_HttpTransportOptions_Builder_setReadTimeout_int_)

## Testing
- Tested the changes by running a pipeline with **GCS Copy** plugin in CDAP sandbox

itsankit-google commented on widgets/GCSMove-action.json:
```diff
@@ -128,6 +128,16 @@
+      "properties" : [
+        {
+          "name": "readTimeout",
+          "widget-type": "textbox",
```
use widget type number, for example: https://github.com/data-integrations/google-cloud/blob/b4c7d043eb84b6611c2d70554233d35cab32ff70/widgets/GCSFile-batchsource.json#L122

This comment applies to whole PR.

dj-smart replied:
Addressed

itsankit-google reviewed (commented):
Please add a validation check for negative integers as well.

Please also include markdown changes.

dj-smart commented:
> Please add a validation check for negative integers as well.
> 
> Please also include markdown changes.

1. For negative integers, it uses the default readTimeout as mentioned here: https://cloud.google.com/java/docs/reference/google-cloud-core/latest/com.google.cloud.http.HttpTransportOptions.Builder#com_google_cloud_http_HttpTransportOptions_Builder_setReadTimeout_int_
2. Added the documentation.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java:
```diff
@@ -128,6 +139,7 @@ public void validate(FailureCollector collector, Map<String, String> arguments)
     collector.getOrThrowException();
   }
 
+
```
Nit: remove empty line.

dj-smart replied:
ACK

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java:
```diff
@@ -74,15 +75,25 @@ public class SourceDestConfig extends GCPConfig {
+  @Description("Timeout in seconds to read data from an established HTTP connection (Default value is 20). " +
+    ("For performing copy/move operation on large files in GCS buckets, set a higher value. " +
+      "Set it to 0 for infinite(no limit)"))
+  protected String readTimeout;
```
If the widget type is number, we can directly read it as Long saves an extra type conversion step.

dj-smart replied:
Set the type to Integer

itsankit-google commented on widgets/GCSMove-action.json:
```diff
@@ -128,6 +128,16 @@
+        {
+          "name": "readTimeout",
+          "widget-type": "number",
+          "label" : "Read Timeout"
```
Can also set default minimum and maximum in the widget. See sample size widget referenced in last comment.

dj-smart replied:
Set the minimum and default values

itsankit-google reviewed (changes requested):
Please also add a validation check if negative value is passed in the config.

itsankit-google commented:
> > Please add a validation check for negative integers as well.
> > Please also include markdown changes.
> 
> 1. For negative integers, it uses the default readTimeout as mentioned here: https://cloud.google.com/java/docs/reference/google-cloud-core/latest/com.google.cloud.http.HttpTransportOptions.Builder#com_google_cloud_http_HttpTransportOptions_Builder_setReadTimeout_int_
> 2. Added the documentation.

I saw that but we could just avoid it by a validation check to imply what is intended.

itsankit-google commented:
Please always add a build label to the PR so that we know build is passing successfully before reviewing, it saves time for reviewers.

dj-smart commented:
> Please always add a build label to the PR so that we know build is passing successfully before reviewing, it saves time for reviewers.

Acknowledged

dj-smart commented:
> > > Please add a validation check for negative integers as well.
> > > Please also include markdown changes.
> > 
> > 
> > 
> > 1. For negative integers, it uses the default readTimeout as mentioned here: https://cloud.google.com/java/docs/reference/google-cloud-core/latest/com.google.cloud.http.HttpTransportOptions.Builder#com_google_cloud_http_HttpTransportOptions_Builder_setReadTimeout_int_
> > 2. Added the documentation.
> 
> I saw that but we could just avoid it by a validation check to imply what is intended.

Added the validation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java:
```diff
@@ -125,9 +136,20 @@ public void validate(FailureCollector collector, Map<String, String> arguments)
+      return;
+    }
+    if (readTimeout < 0) {
+      collector.addFailure("Please enter zero or a positive value"
+        , null).withConfigProperty(READ_TIMEOUT);
```
```
collector.addFailure("Read Timeout cannot be less than 0", "Please enter 0 or positive value").withConfigProperty(READ_TIMEOUT);
```

```Please enter 0 or positive value``` is a corrective action.

dj-smart replied:
Understood. Done.

itsankit-google reviewed (commented):
one minor comment, please squash commits using rebase and test locally.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java:
```diff
@@ -125,9 +136,20 @@ public void validate(FailureCollector collector, Map<String, String> arguments)
+    }
     collector.getOrThrowException();
   }
-
```
why is this removed? There should be an empty line between two methods.

dj-smart replied:
Done

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java:
```diff
@@ -125,9 +136,20 @@ public void validate(FailureCollector collector, Map<String, String> arguments)
+      collector.addFailure("Read Timeout cannot be less than 0",
+                           "Please enter 0 or a positive value").withConfigProperty(READ_TIMEOUT);
+    }
+  }
```
similarly here, add an empty line before another method starts.

dj-smart replied:
Done

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/gcs/StorageClient.java:
```diff
@@ -320,17 +321,21 @@ private static String toPath(BlobId blobId) {
 
-  public static StorageClient create(GCPConnectorConfig config) throws IOException {
-    return create(config.getProject(), config.getServiceAccount(), config.isServiceAccountFilePath());
+  public static StorageClient create(GCPConnectorConfig config, Integer readTimeout) throws IOException {
+    return create(config.getProject(), config.getServiceAccount(), config.isServiceAccountFilePath(), readTimeout);
```
```
public static StorageClient create(GCPConnectorConfig config) throws IOException {
    return create(config.getProject(), config.getServiceAccount(), config.isServiceAccountFilePath(),  null);
```
This way we will not need to change the underlying classes which are calling it and not using `readTimeout`.

dj-smart replied:
Just to clarify, are you proposing to add this additional method along with the existing one?

dj-smart replied:
Discussed offline, made the change
